### PR TITLE
[Feature] Overwrite test by URI during QTI import

### PR DESCRIPTION
**Pull request by shpran**
# [PISA25-774](https://oat-sa.atlassian.net/browse/PISA25-774)

## Changes
* Request parameter `overwriteTest` replaced by `overwriteTestUri`
* Code updated to use `overwriteTestUri` request parameter

# Related PRs

- https://github.com/oat-sa/tao-translation-synchronizer/pull/409

[PISA25-774]: https://oat-sa.atlassian.net/browse/PISA25-774?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: models/classes/import/QtiTestImporter.php [hunk 1]
```diff
@@ -35,32 +38,33 @@ class QtiTestImporter extends AbstractTestImporter
...
-        $enableValidators = true,
-        $itemMustExist = false,
-        $itemMustBeOverwritten = false,
-        bool $overwriteTest = false,
+        core_kernel_classes_Class $class = null,
+        bool $enableMetadataGuardians = true,
+        bool $enableValidators = true,
+        bool $itemMustExist = false,
+        bool $itemMustBeOverwritten = false,
+        ?string $overwriteTestUri = null,
```
Comment (gabrielfs7) [inline models/classes/import/QtiTestImporter.php:57]: Ok, so now we must always know the testUri to be able to use this method and API, which means, we have a breaking change? I am not sure PISA is the only one use it. Would it be possible to no introduce this breaking change and make it backward compatible?

Comment (shpran) [reply models/classes/import/QtiTestImporter.php:57]: As I see, this parameter was added by @yaraslau-kavaliou for PISA. So I think, that no one is using it. @yaraslau-kavaliou can you confirm that?

But sure, for safety I can keep the old one and add this as a new one.

File: models/classes/class.QtiTestService.php [hunk 2]
```diff
@@ -632,7 +632,8 @@ protected function importTest(
                         }
                     }
 
-                    $this->deleteTestsFromClassByLabel($testLabel, $itemsClassLabel, $testClass, $itemParentClass);
+                    $this->getTestService()->deleteTest(new core_kernel_classes_Resource($overwriteTestUri));
```
Comment (gabrielfs7) [inline models/classes/class.QtiTestService.php:635]: - [ ] @shpran We cannot forget to move the test to the new path before deleting it (so we do a clean up).
- [ ] @shpran We cannot forget to move the test items to the new path before deleting them (so we do a clean up).

Comment (shpran) [reply models/classes/class.QtiTestService.php:635]: We provide new path in synchronizer app, so this part already done 🙂 
First - we build new path
Second - we publish the test
During publishing:
First - we delete old one
Second - we publish new one

File: actions/class.RestQtiTests.php [hunk 3]
```diff
@@ -117,9 +126,9 @@ public function import()
                 }
                 return $this->returnSuccess($data);
             } else {
-                throw new \common_exception_RestApi($report->getMessage());
+                throw new common_exception_RestApi($report->getMessage());
             }
-        } catch (\common_exception_RestApi $e) {
+        } catch (common_exception_RestApi $e) {
             return $this->returnFailure($e);
```
Comment (gabrielfs7) [inline actions/class.RestQtiTests.php:132]: @shpran would it be asking too much if we could make sure that the exceptions caught here have logs with: 

- Exception class name that was triggered
- Exception message
- Exception trace

Maybe we already have something on `returnFailure` implementation?

I am asking, cause generally this logs with only exception message do not help at all as we have no idea where it happened. This is a critical process for the whole ingestion process, so I think it is useful to have this logs

Wdyt?

File: actions/class.RestQtiTests.php [hunk 4]
```diff
@@ -191,7 +201,7 @@ public function importDeferred()
             }
 
             return $this->returnSuccess($result);
-        } catch (\common_exception_RestApi $e) {
+        } catch (common_exception_RestApi $e) {
             return $this->returnFailure($e);
```
Comment (gabrielfs7) [inline actions/class.RestQtiTests.php:205]: Same as here https://github.com/oat-sa/extension-tao-testqti/pull/2491/files#r1703668338

File: actions/class.RestQtiTests.php [hunk 5]
```diff
@@ -295,16 +305,37 @@ private function getUploadedPackageData()
...
+        }
+
+        return $overwriteTestUri;
+    }
+
     /**
      * Get class instance to import test
-     * @throws \common_exception_RestApi
      * @return \core_kernel_classes_Class
+     *@throws common_exception_RestApi
```
Comment (gabrielfs7) [inline actions/class.RestQtiTests.php:338]:
```suggestion
     * @throws common_exception_RestApi
```

File: actions/class.RestQtiTests.php (hunk 3 above)
Comment (shpran) [reply actions/class.RestQtiTests.php:132]: This one doesn't relate to my changes (just fixed FQN), but sure I will improve these logs 😎 
Thanks for suggestion

File: models/classes/class.CrudQtiTestsService.php [hunk 6]
```diff
@@ -91,8 +92,15 @@ public function importQtiTest(
...
+                $uploadedFile,
+                $overwriteTest,
+                $itemClassUri,
+                $overwriteTestUri
+            );
             helpers_TimeOutHelper::reset();
+
             return $report;
         } catch (common_exception_UserReadableException $e) {
             return new common_report_Report(common_report_Report::TYPE_ERROR, __($e->getUserMessage()));
```
Comment (gabrielfs7) [inline models/classes/class.CrudQtiTestsService.php:106]: @shpran Can you please add a proper log here like explained here?

https://github.com/oat-sa/extension-tao-testqti/pull/2491/files#r1703668338

Thank you!

File: models/classes/import/QtiTestImporter.php [hunk 7]
```diff
@@ -35,33 +38,37 @@ class QtiTestImporter extends AbstractTestImporter
...
-        \core_kernel_classes_Class $class = null,
-        $enableMetadataGuardians = true,
-        $enableValidators = true,
-        $itemMustExist = false,
-        $itemMustBeOverwritten = false,
+        core_kernel_classes_Class $class = null,
+        bool $enableMetadataGuardians = true,
+        bool $enableValidators = true,
+        bool $itemMustExist = false,
+        bool $itemMustBeOverwritten = false,
```
Comment (gabrielfs7) [inline models/classes/import/QtiTestImporter.php:57]: Are we sure that all places pass proper booleans here? If someone is passing 0 or 1 instead or any other invalid value, we would be breaking compatibility.

File: actions/class.RestQtiTests.php [hunk 8]
```diff
@@ -295,16 +305,37 @@ private function getUploadedPackageData()
...
+            return null;
+        }
+
+        if (!is_string($overwriteTestUri)) {
+            throw new common_exception_RestApi(
+                sprintf('%s parameter should be string', self::OVERWRITE_TEST_URI)
+            );
+        }
+
+        return $overwriteTestUri;
```
Comment (Nevermind23) [inline actions/class.RestQtiTests.php:332]:
```suggestion
        if (!is_null($overwriteTestUri) && !is_string($overwriteTestUri)) {
            throw new common_exception_RestApi(
                sprintf('%s parameter should be string', self::OVERWRITE_TEST_URI)
            );
        }

        return $overwriteTestUri;
```

File: actions/class.RestQtiTests.php (hunk 3 above)
Comment (gabrielfs7) [reply actions/class.RestQtiTests.php:133]: @shpran I know this is not part of your changes, but today is a bit of a pain to track errors when we are generating translations. Could you please add logs for that?

**Review by gabrielfs7 — APPROVED** [review]
Thanks @shpran 

- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

File: actions/class.RestQtiTests.php (hunk 3 above)
Comment (shpran) [reply actions/class.RestQtiTests.php:135]: I checked `returnFailuer` implementation and found, that we use Advanced Logger feature there, so message, file, line and trace should be logged:
image

File: actions/class.RestQtiTests.php (hunk 4 above)
Comment (shpran) [reply actions/class.RestQtiTests.php:220]: https://github.com/oat-sa/extension-tao-testqti/pull/2491#discussion_r1709441560

File: models/classes/class.CrudQtiTestsService.php (hunk 6 above)
Comment (shpran) [reply models/classes/class.CrudQtiTestsService.php:114]: https://github.com/oat-sa/extension-tao-testqti/pull/2491/commits/f6c647fbb273b52b4e0e2b72399a6812444856da

**Pull request merged**